How to create cmdlet w/dotnet CLI

rkeithhill opened the pull request:
Partially addresses #4894 and #4809.

rkeithhill committed:
How to create cmdlet w/dotnet CLI

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+
+## Prerequisites
+* Download and install the [.NET Core 2.0 SDK][net-core-sdk] for your operating system.
+  On Linux, I recommend using a package manager to install the SDK.
```
Probably change to `It is recommended to use a package manager to install the SDK on Linux` or something similar. Documentation should not use first person.

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   ```
+
+5. Load the binary and invoke the new command:
+   ```
````
add `powershell` formatting.

rkeithhill replied:
fixed

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+1. Copy `mymodule.dll` to a folder on a Windows machine.
+
+2. Import the module:
+   ```
````
add `powershell` formatting

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   Note: the module should load without errors.
+
+3. Execute the Write-TimestampedMessage command:
+   ```
````
Add `powershell` formatting

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+If you install the .NET Core 2.0 SDK on Windows, you
+can find the implementation version of netstandard.dll for .NET 4.6.1 in the following directory:
+`C:\Program Files\dotnet\sdk\2.0.0\Microsoft\Microsoft.NET.Build.Extensions\net461\lib`.
+If you copy the netstandard.dll from this directory to the directory containing
```
code escape  `netstandard.dll`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+1. Start a new Windows PowerShell console. Remember that once a binary assembly is loaded into
+   PowerShell it can't be unloaded. Restarting PowerShell is necessary to get it to reload mymodule.dll.
+
+2. Copy the implementation version of netstandard.dll for .NET 4.6.1 to our module directory:
```
code escape `netstandard.dll`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   PowerShell it can't be unloaded. Restarting PowerShell is necessary to get it to reload mymodule.dll.
+
+2. Copy the implementation version of netstandard.dll for .NET 4.6.1 to our module directory:
+   ```
````
add `powershell` formatting

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   ```
+
+3. Import the module and execute the command:
+   ```
````
add `powershell` formatting

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+   This time the command should succeed.
+
+If you use additional libraries there may be more work involved. This approach has
+been successfully tested using types from System.Xml and System.Web.
```
code escape `System.Xml` and `System.Web`.

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+## .NET 4.7.1
+Early testing indicates that Windows systems with .NET 4.7.1 installed or Windows 10
+Fall Creators Update and higher, will load the binary module without the need to have
+netstandard.dll copied beside the module's assembly.
```
code escape  `netstandard.dll`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+
+This example shows how you can create a simple, binary module using the dotnet
+CLI (command line interface).  For this example, you can use any OS that the
+.NET Core 2.0 SDK installs on.  The Windows for Subsystem Linux console works
```
Should be `Windows Subsystem for Linux` not `Windows for Subsystem Linux`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+# Creating a Simple Cross-Platform Binary Module with the Dotnet CLI
+
+This example shows how you can create a simple, binary module using the dotnet
+CLI (command line interface).  For this example, you can use any OS that the
```
my preference is to Use the full name and then parenthesize the initialism `Command Line Interface (CLI)` when bring an initialism into context.

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   dotnet add package PowerShellStandard.Library --version 3.0.0-preview-01
+   ```
+
+3. Add the source code for a PowerShell command by opening the Class1.cs file in an editor and adding the following code:
````
Code escape `Class1.cs`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+can find the implementation version of netstandard.dll for .NET 4.6.1 in the following directory:
+`C:\Program Files\dotnet\sdk\2.0.0\Microsoft\Microsoft.NET.Build.Extensions\net461\lib`.
+If you copy the netstandard.dll from this directory to the directory containing
+mymodule.dll, the `Write-TimestampedMessage` command will work.  Let's try that.
```
code escape `mymodule.dll`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+   ```
+
+This error indicates that the mymodule.dll assembly can't find the implementation version of the
+netstandard.dll for the .NET Framework.
````
code escape `netstandard.dll`

rkeithhill replied:
fixed

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,135 @@
+       + FullyQualifiedErrorId : System.IO.FileNotFoundException
+   ```
+
+This error indicates that the mymodule.dll assembly can't find the implementation version of the
````
code escape `mymodule.dll`

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,135 @@
+mymodule.dll, the `Write-TimestampedMessage` command will work.  Let's try that.
+
+1. Start a new Windows PowerShell console. Remember that once a binary assembly is loaded into
+   PowerShell it can't be unloaded. Restarting PowerShell is necessary to get it to reload mymodule.dll.
```
code escape `mymodule.dll`

rkeithhill replied:
fixed

markekraus reviewed (changes requested):
Thank you! This is very helpful and answers some questions I personally had. There are several things that need to be updated and also a few suggestions.

markekraus commented:
It looks like you may want to code escape occurrences of `dotnet` as well including in the first line.

rkeithhill commented:
@markekraus RE code escpaping `dotnet` - done.  PR update coming soon.

rkeithhill committed:
More editing and address feedback

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -1,33 +1,64 @@
+   ```
+
+3. Add the [PowerShell Standard Library][ps-stdlib] package to the project file.
+   This package provides the System.Management.Automation assembly.
````
Code escape `System.Management.Automation`

daxian-dbw commented:
/cc @anmenaga This PR has some commonalities with #5096.
@rkeithhill You used the PR number in the PR description. Can you please correct it to the corresponding issue number?

rkeithhill committed:
More fixes inlcuding esc S.M.A

rkeithhill commented:
@daxian-dbw RE PR #  - fixed.

rkeithhill commented:
@markekraus RE using `powershell` formatting, can you look at the MD in `view` mode.  It seems GitHub doesn't do a particularly great job at syntax highlighting PowerShell script.  Not the command to add the PowerShellStandard.Library package - the version number is highlighted wrong.  Ditto with the command to add a global.json file and:
```powershell
cd bin/Debug/netstandard2.0
```

rkeithhill committed:
More editing

markekraus commented:
@rkeithhill that's because it should be `Set-Location` 😉 

for the ones that are calling non-PowerShell commands (like `dotnet`), you could try bash syntax. for the ones that are heavy PowerShell some syntax highlighting is better than none. But you could always make them "proper" powershell by avoiding aliases and quoting strings

```powershell
Set-Location "bin/Debug/netstandard2.0"
Import-Module "./mymodule.dll"
Write-TimestampedMessage "Test message"
```

or split them out:
(bash)
```bash
cd bin/Debug/netstandard2.0
```
(pwsh)
```powershell
Import-Module ./mymodule.dll
Write-TimestampedMessage "Test message"
```

rkeithhill commented:
@markekraus No one but N00bs quote paths unless there's a space in the path or they tab-complete a path with spaces.  And "some" power users will escape a space if there's only one instead of quoting the path.  :-)  Ditto for `Set-Location`, most folks use the `cd` alias interactively.

markekraus commented:
@rkeithhill well.. that's completely inaccurate, but opinions on that aside...

I don't think the syntax highlighting is that bad and one reason you want that in is that GitHub is not the only venue these are viewed. In VS Code you get AMAZING syntax highlighting for these. So if it looks a little iffy on GitHub that's fine.

rkeithhill commented:
Yeah, VSCode is where I'm editing this doc.  BTW thanks for all your feedback.

rkeithhill committed:
fix borked mymodule.dd

rkeithhill committed:
Add dotnet --version check

Got burned testing on a machine that had 1.0.1 installed

rkeithhill committed:
Add links to netstandard docs/channel

rkeithhill committed:
Clean up Note: usage

Add  info on restarting Windows PoweShell if command fails.

rkeithhill committed:
Add section header for missing netstandard.dll fix

rkeithhill commented:
Sorry guys!  I've been furiously editing the doc as I test the procedure on other systems.  I'm "done" for a while.  I'll let the requested changes batch up a bit before making any more changes.  And yeah, I'll rebase this PR as it gets closer to final.

rkeithhill commented:
Request for feedback.  I'm a bit inconsistent between using the term `cross-platform` and `portable`.  Cross-platform implies that binary module loads across OS platforms to most folks.  That can be satisfied by compiling against netcoreapp2.0.  Such a binary will load into PS Core running on any platform but not Windows PowerShell.  I guess `portable` is somewhat synonymous.  What I'm really looking for is that word that says "single binary that runs anywhere including in Windows PowerShell".

rkeithhill commented:
After reading through this [dotnet standard issue](https://github.com/dotnet/standard/issues/481) discussion RE problems with loading .NET Standard 2.0 libraries in .NET 4.6.1 applications, I'm think perhaps I should just remove the whole `The fix for missing netstandard.dll`.  I'm probably just getting lucky and this could all fall apart as different netstandard2.0 binary modules that user varying versions of the same package are imported.  Unless and until, Microsoft provides an update to Windows PowerShell 3 and 5.x that provides all the necessary binding redirects (if this is even possible given that Win PS has to deal with add-ins i.e binary modules), maybe our position should be - this works on .NET 4.7.1 and that's that.  IOW your users will have to be on Windows 10 FCU or higher or they will have to update to .NET 4.7.1.  If that isn't acceptable, you'll have to multi-target and build for net461 as well as netstandard2.0.  I guess that procedure could be the subject of another "how to".

rkeithhill committed:
Add doc region to .spelling, fix spelling errors

rkeithhill commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,211 @@
+
+Because the binary module's assembly will be created as a `.NET Standard 2.0 class library`,
+the same assembly can be imported into both PowerShell Core and Windows PowerShell.
+This means you do not have to build separate assemblies to target these two different
```
Going to reword to `build and distribute separate assemblies that target these two
different implementations of PowerShell.`

rkeithhill commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,211 @@
+   ```
+
+   This should output `2.0.0` or higher. If it returns a major version of 1, make sure you have
+   installed the .NET Core 2.0 SDK and have re-started your shell to get the newer version of
````
should be `restarted` with no hypen.

rkeithhill committed:
Add wrapup section and more editing

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,224 @@
+
+## Create the .NET Standard 2.0 Binary Module
+
+1. Verify you are running the 2.0.0 version of the dotnet CLI.
```
I think this should be used for all occurrence: `dotnet` CLI

rkeithhill replied:
Are you saying to code escape dotnet .e.g. `dotnet` everywhere it is used in conjunction with CLI? If so, I can do that.

BTW I had more stuff code-escaped at one point but then a I was reading it, the escaping started to look too busy.  I found I was using it for "code" and to draw attention particular nouns for those who skim.

markekraus replied:
I use code escaping for commands, files names, and assembly names as a general rule. Other things like the SDK name might be a bit much. for this, `dotnet` is a command and "CLI" is a description of that command. I think "`dotnet` CLI" is probably the best approach.

markekraus commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,224 @@
+   (the default for classlib projects).
+
+   ```powershell
+   dotnet new classlib --name mymodule
````
Should `mymodule` be Pascal case? `MyModule` The other occurrences would need to be updated too such as `mymodule.dll` -> `MyModule.dll`

rkeithhill replied:
Good question. The reason I did it that way was because I developed this module on Linux and using all lower case felt more natural.  Perhaps it should be kebab case e.g. my-module.dll (similar to the posh-git module).  Or I can change it to Pascal case.  Let's let a few more folks weigh in.

markekraus replied:
I think the [.NET naming and casing conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions) should apply regardless of OS and case sensitivity (or lack there of) for consistency.

rkeithhill committed:
Fix spelling error

Address feedback on code escaping dotnet.
Get rid of unnecessary code escaping.

rkeithhill committed:
Missed a few changes

rkeithhill committed:
Add to markdown.tests.ps1

Address code block should have language feedback.
Address mymodule should be MyModule feedback.

rkeithhill commented on docs/cmdlet-example/command-line-simple-example.md:
```diff
@@ -0,0 +1,224 @@
+
+## Create the .NET Standard 2.0 Binary Module
+
+1. Verify you are running the 2.0.0 version of the `dotnet` CLI.
```
Done

rkeithhill committed:
Fix some markdown lint issues

Could not resolve any of the MD029 warnings.  From what I can tell , the markdown has a properly incrementing ordered list starting at one.

markekraus reviewed (approved):
LGTM

rkeithhill commented:
Are you guys OK with me removing this md file from the markdown.test.ps1 Pester test?  I fixed all of the rule errors except for the MD029 errors.  Those appear to me to be bogus errors (every list starts with 1 and increments correctly).  Either that or perhaps there is a way to suppress this rule?

markekraus commented:
I think we are using the "One" format. so every item in the list should begin with `1.`

```none
1. Do this.
1. Do that.
1. Done.
```
I think that is actually better long term as it makes it easier to insert items and reorder steps.

daxian-dbw commented on docs/cmdlet-example/command-line-simple-example.md:
````diff
@@ -0,0 +1,227 @@
+       + FullyQualifiedErrorId : System.IO.FileNotFoundException
+   ```
+
+If the command worked, congratulations! You're system was probably updated to
````
> You're system was probably updated to

A typo. Should be **Your**

daxian-dbw commented:
@markekraus is right, we are using "One" style. See https://github.com/PowerShell/PowerShell/blob/master/.markdownlint.json

rkeithhill committed:
Fix grammar error

drewjanderson reviewed (approved):
Well done, have to get that grammar right!

rkeithhill commented:
I'm a little perplexed by an "ordered" list where every item is `1.`.  In that case, what's the point of using an ordered list? 

I'd rather change it an unordered list.  There is a similar [PR](https://github.com/anmenaga/PowerShell/blob/9a0dd5b2f4220dcb167f69d50b26c46cf2f60321/docs/cmdlet-example/README.md) for doing this with Visual Studio.  It uses an ordered list in the form of `1.`, `2.`, `3.`, etc.  How is that documentation getting past this issue?  Hint: it has not been added to the markdown.test.ps1 file.

daxian-dbw commented:
We decided to use the 'One' style because it's easier to add items in the middle of the list in later edits. It will still show as `1,2,3,...` after the markdown file is rendered.

Good point on #5096, I will leave a comment there to get that file added to the markdown test.

rkeithhill commented:
OK the following ordered list is from https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md

![image](https://user-images.githubusercontent.com/5177512/31701523-32844a60-b38e-11e7-9c40-411fb8581e4f.png)

So this MD029 rule (set to One) says that list is not right and each item should start with `1.`, right?

rkeithhill commented:
And this list from the Community Governance md is wrong too:

![image](https://user-images.githubusercontent.com/5177512/31701604-b260bda4-b38e-11e7-8c67-c00012d2536e.png)

daxian-dbw commented on test/common/markdown/markdown.tests.ps1:
```diff
@@ -74,6 +74,7 @@ Describe 'Common Tests - Validate Markdown Files' -Tag 'CI' {
             $docsToTest = @(
                 './*.md'
                 './docs/*.md'
+                './docs/cmdlet-example/command-line-simple-example.md'
```
Maybe we can change this to `'./docs/cmdlet-example/*.md'`, so that #5096 doesn't need to edit `markdown.tests.ps1` again.

daxian-dbw commented:
Those 2 markdown files are not captured in `markdown.test.ps1`, and there are more others too. That test was added after many markdown files were already checked in. Not all `.md` files were added to the `docsToTest` list initially because that would be too many files to fix. The idea was to gradually fix all existing `.md` files, but for new files, make sure they are covered by the test when getting added. /cc @TravisEz13

rkeithhill committed:
Switch olist to One style

rkeithhill commented:
@TravisEz13 OK I think I finally get it the MD029 rule.  While the procedure appears as step 1, step 1, step 1 in the MD file (which looks kinda borked), it actually shows up as step 1, step 2, step 3 in the rendered view.  I was thinking that the rendered view would reflect the step 1, step 1, step 1 - which made no sense to me.  :-)  I updated the PR so it should pass the markdown pester tests now.

daxian-dbw commented:
@rkeithhill Thanks for the update! The ordered style is still used in two other places in `command-line-simple-example.md`. Can you please update them too?

rkeithhill committed:
Argh, wish VSCode would tell me I have unsaved changes when committing

rkeithhill commented:
Just saw that and pushed an update.  One beef I have with VSCode is that it doesn't warn you when you commit a file that you have unsaved changes to the file.  See https://github.com/Microsoft/vscode/issues/33004
